Add `forEachFilter`

0pg opened the pull request:
- `CovariantFilter` typeclass that provides `mapFilter` functionality and has a `Covariant` typeclass
- `ForEachFilter extends CovariantFilter` typeclass that provides `forEachFilter` functionality and has a `ForEach` typeclass

And the `collect` function is following by this work (https://github.com/zio/zio-prelude/issues/629)

adamgfraser commented:
This doesn't make sense. If I remove your overridden implementation for `List` I get the following:

```scala
object Example extends App {

  val as = List(1, 2, 3, 4, 5)

  val filtered = as.forEachFilter(a => if (a % 2 == 0) Id(Some(a)) else Id(None))

  println(filtered) // List()
}
```

More generally, the contract of `ForEach` is that it allows iterating over a structure and reconstructing a structure of exactly the same "shape". Filtering violates this contract since the structure has a different shape if any elements are filtered out. Furthermore, allowing elements to be filtered out implies that potentially all the elements could be filtered out, which is inconsistent with the contract of `NonEmptyForEach` and every `NonEmptyForEach` must by definition be a `ForEach`.

0pg commented:
Ah I understood. then `ForEachFilter` should be separated from `ForEach` 🤔

adamgfraser commented:
Yes it would need to be a new functional abstraction if we want to support it at all.

0pg committed:
Add functional abstractions for filter

Fix comment

Fix comment

Remove unnecessary code

Fix ForEachFilter constructors

Fix ForEachFilter implicit constructors

0pg committed:
Fix broken compile

0pg commented:
`CovariantFilter` provides combined operator that `map` and `filter` conceptually.
But I think that `CovariantFilter` itself cannot be considered as following `Covariant`'s (neither `Invariant`'s) contract so I made it has a `Covariant` (like `cats` does)
What do you think? @adamgfraser

adamgfraser commented:
I think at this point we are adding a functional abstraction because we want to implement a particular operator instead of because we have a deep understanding of the underlying domain, which leads to a proliferation of functional abstractions and a lack of orthogonality.

This operator is conflating a bunch of different things between mapping, iterating, and filtering. We can instead use our existing `IdentityEither` functional abstraction, which describes a `none` value for a data type, and conceptualize filtering as replacing each `A` value with either `succeed` if the predicate is satisfied or the `none` value otherwise. Then it is easy to filter the results after applying any transformation.

See #1138.

0pg commented:
> because we have a deep understanding of the underlying domain, which leads to a proliferation of functional abstractions and a lack of orthogonality.

I was not sure about the filter functionality really needs `Flatten` and `IdentityEither` abstractions, but as far as I understand this sentence says what we actually aims to do
It seems that most of functionality can be given just by composition of already existing functional abstractions

0pg commented:
with the approach using `IdentityEither`, we can say that we already have `ForEachFilter` functionality but it works quite inefficient way.
I have no idea there is a way to solve this problem without a new functional abstraction

adamgfraser commented:
@0pg Going to close this and I will work on optimization separately.
